### PR TITLE
[Review]Acid zombies

### PR DESCRIPTION
**Pull request by PropaneSoup**
This PR adds three new acid based zombies:

-acidic zombie: Toned down spitter. Can't actually spit, but is still quick and leaves an acid puddle on death. Upgrades quickly into either a spitter or glowing.

-glowing zombie: A quick zombie, but not very dangerous by itself. It's main purpose is to find you during a night raid and light up your location.

-corrosive zombie: spitters can upgrade into this. It's rather tanky, with good armor and a lot of health. It's also rather slow, but shoots out acid at twice the rate of a spitter, and leaks acid when it walks and when it gets hit. So taking it on in melee is very dangerous, and you'll always need to watch where you're standing.

### CODE REVIEW

**Comment by Snaaty** [comment]
I really like the way these are looking!

**Comment by Frost-wood** [comment]
looks relatively balanced, with a cool concept.

**Comment by Coolthulhu** [comment]
The weaker spitter zombie should have something to actually do. At least vomit acid on the player when in melee.

Glowing shouldn't be acid based. It simply doesn't fit there, except for video game "acid glows" effect. It would be better with boomer (chemical, affects senses), shocker ("underdeveloped" shocker that only glows when it can melee its target) or shrieker (the other "alert" zombie).

As for corrosive: passive acid leaking is simply a bad idea for a zombie. It would kill all its allies. No other zombie has a passive anti-zed effect and spitters are anti-zombie enough already.
On-hit effect would be OK if it was more of a splash-back thing rather than splash-around. For example, having it splash in the direction of attacker rather than all around.
Also not sure about spamminess. It's basically two spitters merged together with an on-hit effect and OK melee.

The idea I had for a spitter evolution would be giving it a "glob" projectile attack it would use at range every other turn. Something like this would be better than just lots of spit fields, because it couldn't be duplicated by 2x spitter + few other zeds to work as a meat wall.

Finally: code quality. Format it a bit. Use `pos3()` instead of `pos()` where possible - it will make the conversion to 3d faster.

**Comment by Rivet-the-Zombie** [comment]
>Glowing shouldn't be acid based. It simply doesn't fit there, except for video game "acid glows" effect. **It would be better with boomer** (chemical, affects senses)

I'd like to second this.

**Comment by Zireael07** [comment]
+1 to moving glowing to boomer.

**Comment by PropaneSoup** [comment]
> On-hit effect would be OK if it was more of a splash-back thing rather than splash-around. For example, having it splash in the direction of attacker rather than all around.

So, basically the same as zapback, but acid instead of electricity? I did that originally, but I favored creating acid fields simply for the variety. I suppose because of every other zombies vulnerability to acid, it would be better from a game play standpoint to have a zombie that doesn't actively kill its allies, intentionally or otherwise :P 

> giving it a "glob" projectile attack it would use at range every other turn.

So something like an actual projectile that it would "shoot" at you? Perhaps a little brunt damage when it, and then apply a "covered in acid" DoT effect?

> Use pos3() instead of pos()

Will do gladly, anything to help in your heroic undertaking of making this thing 3D ;)

> type of heavily irradiated zombie

I had an idea for a zombie that would have a constant radioactive field around itself, like smokers, but I wasn't sure where to tie it in with the lore/existing zombie types. 

As a side note, brainstorming/approving my ideas would be much more efficient to make a thread in the forums before typing it all up, wouldn't it?

**Comment by DavidKeaton** [comment]
thread in forums would help get videos from the forums only peeps. :-)

**Comment by Coolthulhu** [comment]
> Perhaps a little brunt damage when it, and then apply a "covered in acid" DoT effect?

Just acid damage + maybe small density field on the point that was hit. You could add an ammo effect that would bump current tile's acid field by 1 on hit. Leave DoTs for later (next form or something) - we don't have a good way of removing those, so they could easily get out of hand.

> So, basically the same as zapback, but acid instead of electricity?

I meant you could still just spawn acid fields, but move the center towards the attacker. So that acid can't spawn behind zed's back, but if the player is in melee, it can spawn behind player's back.
As for ranged attacks, you could imagine bullets entering with a large splash, then melting before they can reach the exit wound.
This would make it more of a defense rather than just on-hit effect, while still retaining the difference from zapback.

> heavily irradiated zombie that glows Hollywood-style. Maybe it'd chase you down and barf glowing nuclear sludge

That could be cool - it could be the alternate route from the regular acid barfer and could later have a common advanced form with boomer (fat irradiated glower exploding in radioactive glowing sludge).

**Comment by PropaneSoup** [comment]
With these last two commits; the acidic zombie gets a special acid barf attack and the corrosive zombie gets some alterations to its acid splash-back and its attack. 

For the splash-back, I had to add in a separate check if it was a projectile attack, as I wasn't sure about the vector math that I was going to need to do to have acid always spawn in the direction of the attacker. As for its special attack, it does it every three moves, and its a fairly accurate single 'bolt' of acid. Low damage simply because of how fast it shoots them out.

When testing out these zombies, the biggest thing to fear isn't the raw damage, or the puddles of acid all around you if you run up and start wailing on them, its the sheer amount of pain your character will accumulate by doing it.

**Comment by Coolthulhu** [comment]
For the splash direction, you could do something like this: first define a `sgn(x)` function returning whether `x` is positive, negative or 0, then do something like:
`int dx = sgn( attacker->posx() - zed->posx() )`
Same for `dy`.

Then the field location would be: `zed->posx() + dx * rng(0, 2)` and ditto for the y coord.

For acid bolt, you could grab the projectile code from `spit_sap`. It has the advantage of taking range into account - with your code, the accurate bolt gets more accurate the further you are from the zed due to the way random projectile offset works.

And finally, code quality. You may delay it until you the final moment when all the mechanics are all completed, but it has to be done. Bonus points for using `pos3()` in distance and line calculations.

**Comment by PropaneSoup** [comment]
Alright, I've got the majority of the mechanics worked out, now for balancing ;)

Acid is now splashing in all the right directions, and it looks pretty good. The acid bolt is firing fast and accurate, and I've used `pos3()` wherever the compiler wouldn't yell at me for it. 

In regards to code quality, it all looks consistent with the surrounding code, if I'm wrong on something, could you give me an example?

**Comment by Coolthulhu** [comment]
OK, I tested the zeds.

Barfer is a bit on the weak side, but solid mechanically. Could get a lower cooldown on the ability and leave a density 1 field on the target. Could lose a bit of speed though.
Also, more sane upgrade time - the current one looks like something used to test zed evolution.

As for corrosive: its speed is a bit too low. Makes it too easy to kite away from other zeds.
The acid spreading formula needs tweaking. Current one doesn't work properly when either dx or dy is 0 - the acid then spreads in line.
Maybe something like `z->posx() + dx + rng(-1, 1)`.

Code quality is OK now.
You could use tripoints in acid spreading formula. I think wrapping the variables in `{ }` should be enough for the compiler to recognize it as a tripoint. For example, `g->m.add_field( { x, y, z }, fd_acid, 3, 0 )`.

Balancing on zeds is mostly OK, though. The only major things are the splashback placement with 0 dx or 0 dy and the barfer upgrade time.

**Comment by PropaneSoup** [comment]
Suggested changes are all in, except for the tripoint fields. The only one not using them are the ones in `mondefence`, and the compiler doesn't like the brackets. Other than that, this should be good.

File: src/mondefense.cpp [hunk 1]
```diff
@@ -41,3 +42,40 @@ void mdefense::zapback(monster *const m, Creature *const source, projectile cons
...
+{
+    if( (proj) && ((rng(0, 40) > m->def_chance)) ) {
+        return; //Less likely for a projectile to deliver enough force
+    }
+    int dx = posp( source->posx() - m->posx() );
+    int dy = posp( source->posy() - m->posy() );
+    if (one_in(2)){
+        for (int i = 0; i < rng(3,5); i++) {
+            if (one_in(2)){
+                g->m.add_field(m->posx() + dx + rng(-1, 2),m->posy() + dy + rng(-1, 2), fd_acid, rng(2,3));
```
Comment (Coolthulhu) [inline src/mondefense.cpp]: Should probably be rng(-2, 2), or else the splash will not be symmetrical.

Comment (PropaneSoup) [inline src/mondefense.cpp]: I didn't intend for the larger chance to be symmetrical, I wanted it to be able to go behind the player if they were in melee range. I just realized though that that doesn't accomplish that. Doing `posx() + (dx *2) + rng(-1,1)` would be better

**Comment by Coolthulhu** [comment]
Got a crash during barfer testing. I think it was the blinding - it rolls a random bodypart and applies blinding if it happens to hit the eyes. It applies the status to `foe`, which may be null, which results in a crash.

File: src/monattack.cpp [hunk 2]
```diff
@@ -235,6 +235,154 @@ void mattack::acid(monster *z, int index)
...
+        } else if( seen ) {
+            add_msg( _("The %s barfs acid at %s, but misses!"), z->name().c_str(), target->disp_name().c_str() );
+        }
+        return;
+    }
+
+    body_part hit = random_body_part();
+    int dam = rng(10, 15);
+    dam = target->deal_damage( z, hit, damage_instance( DT_ACID, dam ) ).total_damage();
+    g->m.add_field( foe->pos3(), fd_acid, 1, 0 );
```
Comment (Coolthulhu) [inline src/monattack.cpp]: Here: `foe->pos()`. Foe can be null.

Comment (PropaneSoup) [inline src/monattack.cpp]: Ah, good catch. Changing it now.

**Comment by Coolthulhu** [comment]
Can't get the zeds to evolve properly. Neither acidic->spitter, nor spitter->corrosive.

**Comment by PropaneSoup** [comment]
Oh, it's `upgrade_into` in the json, it should be `upgrades_into`. I'm pretty sure all the other entries are correct.

**Comment by Coolthulhu** [comment]
Acidic zeds upgrade quite fast (within few days) and they're cool enough to have a larger "window of existence". Later you could give them a larger half-life.

**Pull request merged**